### PR TITLE
Understand some inductive cumulativity in old unification 

### LINKED ISSUE #14734: TC search/unification sometimes fails to take universe cumulativity into account
#### Description of the problem

The two modules, differing only in the universe constraints introduced by `Definition Y`, show different behavior during TC search in the `Goal ..` proof script(s). To make matters more confusing, the difference disappears if we start the TC search without introducing the variable `A`.

```coq
#[universes(cumulative=yes)]
 Polymorphic Class t@{+s} (S : Type@{s}) (A B : Prop) :=
  T {
      prf : S -> A -> B;
    }.
Monomorphic Inductive X@{s | Set < s} : Type@{s} := x.

#[refine]
 Instance t_X_refl {A} : t X A A := {|prf := _|}.
Proof. auto. Qed.

Module A.
  (* Note universe constraint! *)
  Monomorphic Definition Y@{s | X.s < s} : Type@{s} := X.

  Goal forall A, t Y A A.
    Fail intros; typeclasses eauto. (* fails but should succeed, I think *)
    typeclasses eauto.    (* unexpectedly succeeds *)
  Qed.
End A.


Module B.
  (* No universe constraint this time! *)
  Monomorphic Definition Y@{s} : Type@{s} := X.

  Goal forall A, t Y A A.
    Fail Fail intros; typeclasses eauto. (* succeeds now *)
    typeclasses eauto.                   (* succeeds as before *)
  Qed.
End B.
```

#### Coq Version

8.13.2

### PR DESCRIPTION
**Pull request by SkySkimmer**
Revive https://github.com/coq/coq/pull/12624
Fix https://github.com/coq/coq/issues/14734
Fix (partial?) https://github.com/coq/coq/issues/6976

Overlays: https://github.com/mattam82/Coq-Equations/pull/424 https://github.com/HoTT/HoTT/pull/1574

### CODE REVIEW

**Comment by Alizter** [comment]
@coqbot ci minimize

File: test-suite/bugs/closed/bug_14734.v [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#[universes(cumulative=yes)]
+ Polymorphic Class t@{+s} (S : Type@{s}) (A B : Prop) :=
```
Comment (Alizter) [inline test-suite/bugs/closed/bug_14734.v:2]: Don't we have attributes for these now?

Comment (SkySkimmer) [reply test-suite/bugs/closed/bug_14734.v:2]: Polymorphic is prettier though

**Comment by SkySkimmer** [comment]
ping @coq/pretyper-maintainers

**Review by herbelin — APPROVED** [review]
Beside `Constr.compare_head_gen_leq_with` which is a bit hard to follow (but this is not new), the PR looks good. I'm not a cumulativity expert though.

**Comment by ppedrot** [comment]
I wouldn't mind a high level description of the changes. The major difference is that the equality function is now aware of the arity of inductive types, but this looks hardly enough to really be more complete?

File: engine/eConstr.ml [hunk 2]
```diff
@@ -570,16 +570,16 @@ let test_constr_universes env sigma leq m n =
           Constr.compare_head_gen_leq_with kind kind leq_universes leq_sorts
             eq_constr' leq_constr' nargs m n
         and leq_constr' nargs m n = m == n || compare_leq nargs m n in
-        compare_leq 0 m n
+        compare_leq nargs m n
```
Comment (SkySkimmer) [inline engine/eConstr.ml:573]: The change is that we thread `nargs` down to here.

Comment (ppedrot) [reply engine/eConstr.ml:573]: M'kay. We should do something with these equality functions btw, they're really unreadable.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@SkySkimmer the two overlays are merged, can you rebase and remove the corresponding commit?

**Comment by SkySkimmer** [comment]
Bench to exercise the machines https://gitlab.com/coq/coq/-/jobs/1621327786
